pr_agent/git_providers/bitbucket_server_provider.py

```diff
@@ -0,0 +1,353 @@
+import difflib
+import json
+import re
+from typing import Optional, Tuple
+from urllib.parse import urlparse
+
+import requests
+from atlassian.bitbucket import Bitbucket
+from starlette_context import context
+
+from .git_provider import FilePatchInfo, GitProvider, EDIT_TYPE
+from ..algo.pr_processing import find_line_number_of_relevant_line_in_file
+from ..algo.utils import load_large_diff
+from ..config_loader import get_settings
+from ..log import get_logger
+
+
+class BitbucketServerProvider(GitProvider):
+    def __init__(
+        self, pr_url: Optional[str] = None, incremental: Optional[bool] = False
+    ):
+        s = requests.Session()
+        try:
+            bearer = context.get("bitbucket_bearer_token", None)
+            s.headers["Authorization"] = f"Bearer {bearer}"
+        except KeyError:
+            s.headers[
+                "Authorization"
+            ] = f'Bearer {get_settings().get("BITBUCKET.BEARER_TOKEN", None)}'
+
+        s.headers["Content-Type"] = "application/json"
+        self.headers = s.headers
+        self.bitbucket_server_url = None
+        self.workspace_slug = None
+        self.repo_slug = None
+        self.repo = None
+        self.pr_num = None
+        self.pr = None
+        self.pr_url = pr_url
+        self.temp_comments = []
+        self.incremental = incremental
+        self.diff_files = None
+        self.bitbucket_pull_request_api_url = pr_url
+
+        self.bitbucket_server_url = self._parse_bitbucket_server(url=pr_url)
+        self.bitbucket_client = Bitbucket(url=self.bitbucket_server_url,
+                                          token=get_settings().get("BITBUCKET.BEARER_TOKEN", None))
+
+        if pr_url:
+            self.set_pr(pr_url)
+
+    def get_repo_settings(self):
+        try:
+            url = (f"{self.bitbucket_server_url}/projects/{self.workspace_slug}/repos/{self.repo_slug}/src/"
+                   f"{self.pr.destination_branch}/.pr_agent.toml")
+            response = requests.request("GET", url, headers=self.headers)
+            if response.status_code == 404:  # not found
+                return ""
+            contents = response.text.encode('utf-8')
+            return contents
+        except Exception:
+            return ""
+
+    def publish_code_suggestions(self, code_suggestions: list) -> bool:
```

Comment:
**Suggestion:** Add type hints to the function parameters of `BitbucketServerProvider.publish_code_suggestions()`.
```suggestion
    def publish_code_suggestions(self, code_suggestions: list[dict]) -> bool:
```